Add the HTML element option to columns and column block

carolinan opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Adds the HTML element option to the columns and column blocks.
Closes #33343

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Add a columns block
2. Open the advanced section and change the HTML element
3. Save and view the front. Confirm that the selected HTML element is used.


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Enhancement

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

carolinan committed:
Add HTML element option to columns

carolinan committed:
Add the HTML element option to the column block

Add the HTML element option to the column block.
Update fixtures.

carolinan committed:
Update core__columns__is-not-stacked-on-mobile.json

Mamaduka commented:
Sidenote: Now that four blocks have this option. Should we extract it as a custom component in the block-editor package?

carolinan commented:
> Sidenote: Now that four blocks have this option. Should we extract it as a custom component in the block-editor package?

[> I wonder if it would also make sense to move the HTML filter to its own separate control. It is used but slightly differently for the group, template part and query blocks.](https://github.com/WordPress/gutenberg/issues/33343#issuecomment-929769608)

Possibly. I am not sure I can do that though.

Mamaduka commented:
Sorry, I miss that comment. Happy to do it as a follow-up PR.

carolinan commented:
I ran `npm run fixtures:regenerate test/integration/full-content/full-content.test.js`

But I then get changes like these which I don't believe are correct? I am using Windows...
```
test\integration\fixtures\blocks\core__column.json
-		"originalContent": "<div class=\"wp-block-column\">\n\t\n\t\n</div>"
+		"originalContent": "<div class=\"wp-block-column\">\r\n\t\r\n\t\r\n</div>"
```

Mamaduka commented:
I just tried to regenerate fixtures locally. I get to errors related to ` core__column__deprecated*` files, and fixtures for other blocks also get updated.

Unfortunately, this is all my knowledge of fixture regeneration - run command and hope it works.

carolinan commented:
yes I am seeing two errors for `core__column__deprecated*` too

mtias commented:
Not to block this PR, but I'm curious if we should aim for making Column be just a variation of Group with some layout settings locked and inheriting everything else (html element, background tools, etc). cc @youknowriad

youknowriad commented:
@mtias yes, this makes sense to me, the `Columns` block is nothing more than a container (group) of group blocks with a specific layout. We need to make sure the "columns" blocks is using the "flex" layout, make sure the "width" is supported for group blocks too and see how/if we can migrate.

carolinan committed:
try updating fixtures

Ran npm run fixtures:regenerate test/integration/full-content/full-content.test.js

talldan commented on packages/block-library/src/column/edit.js:
```diff
@@ -20,12 +21,30 @@ import {
+		verticalAlignment,
+		width,
+		templateLock = false,
+		tagName: TagName = 'div',
```
I see the attribute has a default, so I wonder whether the default value is needed here.

talldan commented on packages/block-library/src/columns/edit.js:
```diff
@@ -52,14 +54,33 @@ import {
+	const {
+		isStackedOnMobile,
+		verticalAlignment,
+		tagName: TagName = 'div',
```
Ditto the previous comment here.

talldan reviewed (commented):
The test failure for `core__columns__deprecated.json` looks correct. The file represents block after it has been upgraded (the deprecation has already been handled), so its attributes reflect the current block.json with the new tagName attribute. To get it passing, delete the file and then run the command to regenerate the fixture.

That might also make the other test failure pass (🤞). My guess is that one is failing because it tries to serialize the block from `core__columns__deprecated.json` to HTML, but the `tagName` attribute is `undefined`, which causes the react error.

carolinan committed:
simplify tagName, try to update fixture

carolinan committed:
Merge branch 'trunk' into update/column-html-element-option

carolinan committed:
fix spacing after merge conflict

carolinan commented:
I removed the unnecessary constant but I am not able to solve the problem with the fixtures.
